### PR TITLE
codegen: split gc roots from other bits on stack

### PR DESCRIPTION
**Pull request by vtjnash**
In order to help avoid memory provenance issues, and better utilize stack space (somewhat), and use FCA less, change the preferred representation of an immutable object to be a pair of `<packed-data,roots>` values. This packing requires some care at the boundaries and if the expected field alignment exceeds that of a pointer. The change is expected to eventually make codegen more flexible at representing unions of values with both bits and pointer regions.

Eventually we can also have someone improve the late-gc-lowering pass to take advantage of this increased information accuracy, but currently it will not be any better than before at laying out the frame.

### CODE REVIEW

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your job](https://github.com/JuliaLang/julia/pull/55767#issuecomment-2350106572) failed.

**Comment by gbaraldi** [comment]
I wonder if we can add some tests in llvmpasses for this. Not as much to test but as documentation as to how this is laid out in the IR

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your job](https://github.com/JuliaLang/julia/pull/55767#issuecomment-2350782624) failed.

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/55767#issuecomment-2350795055) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/84ba536_vs_2ee6551/report.md).

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/55767#issuecomment-2350987608) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/17dc1be_vs_2ee6551/report.md).

**Comment by vtjnash** [comment]
This looks like an LLVM bug to me, but might be helpful if someone could confirm it isn't a misunderstanding in our use of attributes. There seems to be a null pointer dereference happening here since SimplifyCFGPass hoisted too many loads over this branch.

Before SimplifyCFGPass
```
 574737 L76:                                              ; preds = %L74, %L63, %L13
 574738   %value_phi = phi i8 [ 1, %L13 ], [ 1, %L63 ], [ 0, %L74 ]          <= %1 contains a nullptr iff this is 1
 574739   %value_phi2 = phi i64 [ %value_phi16, %L74 ], [ undef, %L63 ], [ undef, %L13 ]
 574740   call void @llvm.memcpy.p0.p0.i64(ptr align 8 %2, ptr align 8 %1, i64 8, i1 false), !dbg !53
 574741   br label %L80, !dbg !53       <= eventually branches to L82 iff %1 is 0
...
 574751 L82:                                              ; preds = %L80
 574752   %75 = load ptr addrspace(10), ptr %2, align 8, !dbg !147, !tbaa !11, !alias.scope !15, !noalias !18, !nonnull !0, !dereferenceable !40, !align !41     <= these attributes are true only if this statement is reached
 574753   %76 = addrspacecast ptr addrspace(10) %75 to ptr addrspace(11), !dbg !152
 574754   %77 = getelementptr inbounds i8, ptr addrspace(11) %76, i32 32, !dbg !152
 574755   %78 = load i64, ptr addrspace(11) %77, align 8, !dbg !152, !tbaa !46, !alias.scope !50, !noalias !51    ; jtbaa_mutab
 574756   br label %L85, !dbg !61
```
After SimplifyCFGPass
```
 575017 L76:                                              ; preds = %pass, %L63, %L13
 575018   %value_phi = phi i8 [ 1, %L13 ], [ 1, %L63 ], [ 0, %pass ]
 575019   %value_phi2 = phi i64 [ %value_phi16, %pass ], [ undef, %L63 ], [ undef, %L13 ]
 575020   call void @llvm.memcpy.p0.p0.i64(ptr align 8 %2, ptr align 8 %1, i64 8, i1 false), !dbg !53
 575021   %74 = trunc i8 %value_phi to i1, !dbg !142
 575022   %75 = xor i1 %74, true, !dbg !142
 575023   %76 = load ptr addrspace(10), ptr %2, align 8, !dbg !142, !nonnull !0, !align !41    <= nonnull still here, despite not being true after hoisting, but dereferenceable is gone as expected
 575024   %77 = addrspacecast ptr addrspace(10) %76 to ptr addrspace(11), !dbg !142
 575025   %78 = getelementptr inbounds i8, ptr addrspace(11) %77, i32 32, !dbg !142
 575026   %79 = load i64, ptr addrspace(11) %78, align 8, !dbg !142                                   <= this is null, and not dereferenceable anymore after hoisting the previous load
 575027   %value_phi3 = select i1 %75, i8 0, i8 1, !dbg !142
 575028   %value_phi4 = select i1 %75, i64 %79, i64 undef, !dbg !142
 575029   %80 = trunc i8 %value_phi3 to i1, !dbg !143
 575030   %81 = xor i1 %80, true, !dbg !143
 575031   br i1 %81, label %L101, label %L88, !dbg !143
```

**Comment by gbaraldi** [comment]
What code generates this.

**Comment by vtjnash** [comment]
original code generated with `julia -O0` (turn off SROA) and `code_llvm(collect, (Base.Generator{Base.ValueIterator{Dict{Base.Docs.Binding, Set{Type}}}, typeof(length)},), raw=true, dump_module=true)`: [output.txt](https://github.com/user-attachments/files/17014805/output.txt)

reproduce this with

```
./usr/tools/llvm-diff  output.ll <(./usr/tools/opt -passes='simplifycfg' output.ll -o - -S) 2>&1 | less
```

**Comment by vtjnash** [comment]
LLVM bug report created: https://github.com/llvm/llvm-project/issues/108854. Probably is a few way we could work around this LLVM bug here in the way we generate the IR, but let's see what they say first.

**Comment by vtjnash** [comment]
I did a manual SROA during codegen, so resulting code might be a little bit worse sometimes, but also more visible to our own gc-late-lowering pass, so that could be better (if our pass was better)

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/55767#issuecomment-2357215135) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/21430a1_vs_0073917/report.md).

**Comment by vtjnash** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/55767#issuecomment-2358268306) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/485ae11_vs_2616634/report.html) is available.

**Comment by vtjnash** [comment]
Looks like some integration issues with FunctionWrappers and still a bit of an issue with the emit_vi_assignment_unboxed function (I thought I fixed that locally already), but otherwise good

**Comment by vtjnash** [comment]
@nanosoldier `runtests(["MethodInspector", "TreesHeaps", "UnsafeAtomicsLLVM", "AutoBZCore", "Hecke", "PyMNE", "PDBTools", "ParameterizedQuantumControl", "StateSpaceEcon", "SymbolicUtils", "InfiniteOpt", "Polynomials4ML", "DECAES", "FiniteMPS", "SphericalFunctions", "LiftAndLearn", "ITensorMPS", "FSimBase", "ApproxMasterEqs", "DINCAE", "SurfaceReactions", "SurfaceCoverage", "StirredReactor", "CellularPotts", "PlugFlowReactor", "NeuronBuilder", "MicroCanonicalHMC", "TulipaPlots", "Chamber", "Pioran", "UnfoldMakie"])`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/55767#issuecomment-2367100071) has completed - no new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/92fddc7_vs_2616634/report.html) is available.

File: test/compiler/codegen.jl [hunk 1]
```diff
@@ -501,10 +501,9 @@ function f37262(x)
     end
 end
 @testset "#37262" begin
-    str = "store volatile { i8, {}*, {}*, {}*, {}* } zeroinitializer, { i8, {}*, {}*, {}*, {}* }* %phic"
-    str_opaque = "store volatile { i8, ptr, ptr, ptr, ptr } zeroinitializer, ptr %phic"
+    str_opaque = "getelementptr inbounds i8, ptr %.roots.phic, i32 16\n  store volatile ptr null"
```
Comment (vtjnash) [inline test/compiler/codegen.jl:504]:
```suggestion
    str_opaque = "getelementptr inbounds i8, ptr %.roots.phic, i32 8\n  store volatile ptr null"
```
(happens to be valid on both 32 and 64 bit archs)

File: src/llvm-late-gc-lowering.cpp [hunk 2]
```diff
@@ -1494,18 +1501,17 @@ SmallVector<Value*, 0> ExtractTrackedValues(Value *Src, Type *STy, bool isptr, I
...
+//    auto Ptrs = ExtractTrackedValues(Src, STy, isptr, irbuilder);
+//    for (unsigned i = 0; i < Ptrs.size(); ++i) {
+//        Value *Elem = Ptrs[i];
+//        Value *Slot = irbuilder.CreateConstInBoundsGEP1_32(irbuilder.getInt8Ty(), Dst, i * sizeof(void*));
+//        StoreInst *shadowStore = irbuilder.CreateAlignedStore(Elem, Slot, Align(sizeof(void*)));
+//        shadowStore->setOrdering(AtomicOrdering::NotAtomic);
+//        // TODO: shadowStore->setMetadata(LLVMContext::MD_tbaa, tbaa_gcframe);
+//    }
+//    return Ptrs.size();
+//}
```
Comment (topolarity) [inline src/llvm-late-gc-lowering.cpp:1514]: Why not delete this?

File: src/cgutils.cpp [hunk 3]
```diff
@@ -1078,6 +1098,244 @@ static void emit_memcpy(jl_codectx_t &ctx, Value *dst, jl_aliasinfo_t const &dst
...
+    return std::make_pair(dst_off + fldoff - src_off, -1);
+}
+
+static void recombine_value(jl_codectx_t &ctx, const jl_cgval_t &x, Value *dst, jl_aliasinfo_t const &dst_ai, Align alignment, bool isVolatileStore)
+{
+    jl_datatype_t *typ = (jl_datatype_t*)x.typ;
+    assert(jl_is_concrete_type(x.typ));
+    assert(typ->layout->first_ptr >= 0 && !x.inline_roots.empty());
+    Align align_dst = alignment;
+    Align align_src(1); // TODO: compute this (and keep aligned)
```
Comment (topolarity) [inline src/cgutils.cpp:1300]: Important TODO?

File: src/cgutils.cpp [hunk 4]
```diff
@@ -3854,25 +4169,24 @@ static jl_cgval_t emit_new_struct(jl_codectx_t &ctx, jl_value_t *ty, size_t narg
...
                 init_as_value = true;
             }
-            else if (tracked.count) {
-                init_as_value = true;
-            }
 
             Instruction *promotion_point = nullptr;
             ssize_t promotion_ssa = -1;
             Value *strct;
+            SmallVector<Value*,0> inline_roots;
```
Comment (topolarity) [inline src/cgutils.cpp:4184]: These are more like `outline_roots`, aren't they?

File: src/cgutils.cpp [hunk 5]
```diff
@@ -1078,6 +1098,244 @@ static void emit_memcpy(jl_codectx_t &ctx, Value *dst, jl_aliasinfo_t const &dst
...
+        src_off = ptr + sizeof(void*);
+        if (!nodata) {
+            assert(dst_off + sizeof(void*) == src_off);
+            dst_off = src_off;
+        }
+    }
+    return std::make_pair(dst_off + fldoff - src_off, -1);
+}
+
+static void recombine_value(jl_codectx_t &ctx, const jl_cgval_t &x, Value *dst, jl_aliasinfo_t const &dst_ai, Align alignment, bool isVolatileStore)
```
Comment (topolarity) [inline src/cgutils.cpp:1294]:
```suggestion

// Copy `x` to `dst`, copying any outlined roots to their native location in the struct body
static void recombine_value(jl_codectx_t &ctx, const jl_cgval_t &x, Value *dst, jl_aliasinfo_t const &dst_ai, Align alignment, bool isVolatileStore)
```

File: src/cgutils.cpp [hunk 6]
```diff
@@ -1078,6 +1098,244 @@ static void emit_memcpy(jl_codectx_t &ctx, Value *dst, jl_aliasinfo_t const &dst
     emit_memcpy_llvm(ctx, dst, dst_ai, data_pointer(ctx, src), src_ai, sz, align_dst, align_src, is_volatile);
 }
 
+static bool allpointers(jl_datatype_t *typ)
+{
+    return jl_datatype_size(typ) == typ->layout->npointers * sizeof(void*);
+}
+
+// compute the space required by split_value_into, by simulating it
```
Comment (topolarity) [inline src/cgutils.cpp:1106]:
```suggestion
// compute the space required by split_value_into, by simulating it
// returns (sizeof(split_value), n_pointers)
```

File: src/cgutils.cpp [hunk 7]
```diff
@@ -1078,6 +1098,244 @@ static void emit_memcpy(jl_codectx_t &ctx, Value *dst, jl_aliasinfo_t const &dst
...
+    bool nodata = allpointers(typ);
+    if (nodata)
+        dst_off = 0;
+    else
+        dst_off = jl_datatype_size(typ);
+    return std::make_pair(dst_off, npointers);
+}
+
+// take a value `x` and split its bits into dst and the roots into inline_roots
+static void split_value_into(jl_codectx_t &ctx, const jl_cgval_t &x, Align align_src, Value *dst, jl_aliasinfo_t const &dst_ai, Value *inline_roots_ptr, jl_aliasinfo_t const &roots_ai, bool isVolatileStore=false)
```
Comment (topolarity) [inline src/cgutils.cpp:1131]: Is this code still used? It looks like this is for the "contiguous block" format, which I believe you opted to split into scalars instead?

File: src/cgutils.cpp [hunk 8]
```diff
@@ -1078,6 +1098,244 @@ static void emit_memcpy(jl_codectx_t &ctx, Value *dst, jl_aliasinfo_t const &dst
...
+    size_t src_off = 0;
+    size_t dst_off = 0;
+    assert(typ->layout->first_ptr >= 0);
+    size_t npointers = typ->layout->npointers;
+    bool nodata = allpointers(typ);
+    for (size_t i = 0; i < npointers; i++) {
+        size_t ptr = jl_ptr_offset(typ, i) * sizeof(void*);
+        if (ptr >= fldoff) {
+            if (ptr >= fldoff + jl_field_size(typ, idx))
+                break;
```
Comment (topolarity) [inline src/cgutils.cpp:1280]: When do we hit this `break`, given the `return` below?

File: src/cgutils.cpp [hunk 9]
```diff
@@ -1078,6 +1098,244 @@ static void emit_memcpy(jl_codectx_t &ctx, Value *dst, jl_aliasinfo_t const &dst
...
+    jl_datatype_t *typ = (jl_datatype_t*)x.typ;
+    auto sizes = split_value_size(typ);
+    AllocaInst *bits = sizes.first > 0 ? emit_static_alloca(ctx, sizes.first, Align(julia_alignment((jl_value_t*)typ))) : nullptr;
+    SmallVector<Value*,0> roots(sizes.second);
+    auto stack_ai = jl_aliasinfo_t::fromTBAA(ctx, ctx.tbaa().tbaa_stack);
+    split_value_into(ctx, x, x_alignment, bits, stack_ai, MutableArrayRef(roots));
+    return std::make_pair(bits, roots);
+}
+
+static std::pair<ssize_t,ssize_t> split_value_field(jl_datatype_t *typ, unsigned idx)
```
Comment (topolarity) [inline src/cgutils.cpp:1268]: I tried to write a doc-comment for this, but could not get the meaning precise enough

Can you help?

File: src/codegen.cpp [hunk 10]
```diff
@@ -3484,25 +3538,26 @@ static size_t emit_masked_bits_compare(callback &emit_desc, jl_datatype_t *aty,
 static Value *emit_bits_compare(jl_codectx_t &ctx, jl_cgval_t arg1, jl_cgval_t arg2)
 {
     ++EmittedBitsCompares;
+    jl_value_t *argty = (arg1.constant ? jl_typeof(arg1.constant) : arg1.typ);
```
Comment (topolarity) [inline src/codegen.cpp:3541]: Is this a related change?

File: src/codegen.cpp [hunk 11]
```diff
@@ -5019,26 +5080,18 @@ static jl_cgval_t emit_call_specfun_other(jl_codectx_t &ctx, bool is_opaque_clos
...
-    case jl_returninfo_t::SRet:
-        result = emit_static_alloca(ctx, getAttributeAtIndex(returninfo.attrs, 1, Attribute::StructRet).getValueAsType(), Align(julia_alignment(jlretty)));
-        argvals[idx] = result;
-        idx++;
-        break;
-    case jl_returninfo_t::Union:
-        result = emit_static_alloca(ctx, ArrayType::get(getInt8Ty(ctx.builder.getContext()), returninfo.union_bytes), Align(returninfo.union_align));
+
+    if (returninfo.cc == jl_returninfo_t::SRet || returninfo.cc == jl_returninfo_t::Union) {
+        result = emit_static_alloca(ctx, returninfo.union_bytes, Align(returninfo.union_align));
```
Comment (topolarity) [inline src/codegen.cpp:5085]: Very nice 👍

**Review by topolarity — APPROVED** [review]
I'm quite happy to have a tighter correspondence between our ABI and unboxed representation (i.e. a very similar root split for `return_roots` and on-stack values)

This feels like a good change to me, esp. since pointer-ful code is often limited by boxing / unboxing and this sets us up for a proper version of https://github.com/JuliaLang/julia/pull/55045

A couple minor things for clarity, but otherwise this was pretty easy to follow (given the size)

File: src/llvm-late-gc-lowering.cpp (hunk 2 above)
Comment (vtjnash) [reply src/llvm-late-gc-lowering.cpp:1514]: I wasn't certain if it would remain unused, and there is a lot more code in here that should be deleted now too, but for being able to differentiate whether future bugs come from removing this now-duplicate support code or from the change if representation in this PR

File: src/cgutils.cpp (hunk 3 above)
Comment (vtjnash) [reply src/cgutils.cpp:1300]: No really. It could impact performance to make this better, but is correct as written

File: src/cgutils.cpp (hunk 4 above)
Comment (vtjnash) [reply src/cgutils.cpp:4184]: haha, I suppose so now

File: src/cgutils.cpp (hunk 7 above)
Comment (vtjnash) [reply src/cgutils.cpp:1131]: IIRC, this is used for places like return_roots and arguments, which need the contiguous block format

Comment (topolarity) [reply src/cgutils.cpp:1131]: Ah that makes sense, I guess we're stuck with both for now

File: src/cgutils.cpp (hunk 9 above)
Comment (vtjnash) [reply src/cgutils.cpp:1268]: It returns the offset values corresponding to jl_field_offset into the two buffers for a split value (or -1)

File: src/codegen.cpp (hunk 10 above)
Comment (vtjnash) [reply src/codegen.cpp:3541]: No, you are just seeing https://github.com/JuliaLang/julia/pull/55770 in the diff. Apparently this needs a rebase to drop that from here now

File: src/cgutils.cpp (hunk 9 above)
Comment (topolarity) [reply src/cgutils.cpp:1268]: Ah, that makes a lot of sense. And `-1` I assume just means not present in one or the other

Throw a `//` in front of that and I'm happy 😉

File: src/llvm-late-gc-lowering.cpp (hunk 2 above)
Comment (topolarity) [reply src/llvm-late-gc-lowering.cpp:1514]: Sounds fine to me - might be nice to make a note about when we can delete it though

File: src/cgutils.cpp (hunk 3 above)
Comment (topolarity) [reply src/cgutils.cpp:1300]: No problem then - would be nice to label it a `perf TODO` though

File: src/cgutils.cpp (hunk 8 above)
Comment (topolarity) [reply src/cgutils.cpp:1280]: Ah right, I see it now. Thanks!

**Comment by fredrikekre** [comment]
This seems to have broken SIMD.jl
```
julia> using SIMD
Info Given SIMD was explicitly requested, output will be shown live

[1491404] signal 11 (128): Segmentation fault
in expression starting at /home/fredrik/.julia/packages/SIMD/0q83J/src/precompile.jl:3
Vec at /home/fredrik/.julia/packages/SIMD/0q83J/src/simdvec.jl:2 [inlined]
Vec at /home/fredrik/.julia/packages/SIMD/0q83J/src/simdvec.jl:6
unknown function (ip: 0x7f7f5175ea74)
macro expansion at /home/fredrik/.julia/packages/SIMD/0q83J/src/precompile.jl:5 [inlined]
macro expansion at /home/fredrik/.julia/packages/PrecompileTools/L8A3n/src/workloads.jl:78 [inlined]
top-level scope at /home/fredrik/.julia/packages/SIMD/0q83J/src/precompile.jl:102
jl_toplevel_eval_flex at /home/fredrik/julia-master/src/toplevel.c:1034
jl_toplevel_eval_flex at /home/fredrik/julia-master/src/toplevel.c:983
ijl_toplevel_eval at /home/fredrik/julia-master/src/toplevel.c:1054
ijl_toplevel_eval_in at /home/fredrik/julia-master/src/toplevel.c:1096
eval at ./boot.jl:439 [inlined]
include_string at ./loading.jl:2676
_include at ./loading.jl:2736
include at ./Base.jl:582
jfptr_include_38137 at /home/fredrik/julia-master/usr/lib/julia/sys.so (unknown line)
jl_apply at /home/fredrik/julia-master/src/julia.h:2220 [inlined]
jl_f__call_latest at /home/fredrik/julia-master/src/builtins.c:883
include at /home/fredrik/.julia/packages/SIMD/0q83J/src/SIMD.jl:1
```

**Comment by topolarity** [comment]
Should be fixed by https://github.com/JuliaLang/julia/pull/55880

**Comment by Zentrik** [comment]
Btw you might have missed some performance regressions this caused as they didn't show up in the NanoSoldier Report (the noise tolerance was higher than the regression):
- `[["array", "index", ("sumcartesian_view", "BitMatrix")]]` got 40% slower.
- `[["array", "index", ("sumelt", "SubArray{Float32, 2, Base.ReshapedArray{Float32, 2, SubArray{Float32, 3, BaseBenchmarks.ArrayBenchmarks.ArrayLS{Float32, 3}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}}, false}, Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}, Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}}, Tuple{Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}}, false}")]]` and the `Int32` version got 30% slower

**Comment by vtjnash** [comment]
I thought those two had often been showing a bidmodal pattern, with no clear reason for it, so I assume that is random noise?

**Comment by Zentrik** [comment]
Both slowdowns are consistent for me locally between this commit and its parent.
